File: api/hyper_trader_api/schemas/trader_config.py
"""
Trader configuration schema for HyperTrader API.

Pydantic v2 models matching the HyperTrader JSON config schema.
"""

from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


class SelfAccount(BaseModel):
    """Self trading account configuration."""

    model_config = ConfigDict(extra="forbid")

    address: str = Field(..., pattern=r"^0x[a-fA-F0-9]{40}$")
    is_sub: bool = False


class CopyAccount(BaseModel):
    """Account to copy trades from."""

    model_config = ConfigDict(extra="forbid")

    address: str = Field(..., pattern=r"^0x[a-fA-F0-9]{40}$")


class ProviderRiskParameters(BaseModel):
    """Provider-level risk management parameters."""

    model_config = ConfigDict(extra="forbid")

    allowed_assets: Literal["*"] | list[str] = Field(
        ...,
        description="'*' for all assets, or a non-empty list of asset names",
    )
    blocked_assets: list[str] = Field(default_factory=list)
    max_leverage: int | None = Field(default=None, ge=1, le=50)

    @field_validator("allowed_assets")
    @classmethod
    def _non_empty_list(cls, v: "Literal['*'] | list[str]") -> "Literal['*'] | list[str]":
        if isinstance(v, list) and len(v) == 0:
            raise ValueError("allowed_assets must be '*' or a non-empty list")
        return v


class ProviderSettings(BaseModel):
    """Exchange and account configuration."""

    model_config = ConfigDict(extra="forbid")

    exchange: Literal["hyperliquid"] = "hyperliquid"
    network: Literal["mainnet", "testnet"]
    self_account: SelfAccount
    copy_account: CopyAccount
    slippage_bps: int = Field(default=200, ge=0, le=1000)
    risk_parameters: ProviderRiskParameters


class OpenOnLowPnl(BaseModel):
    """Configuration for opening positions when portfolio PnL is low."""

    model_config = ConfigDict(extra="forbid")

    enabled: bool = False
    max_pnl: float = Field(default=0.0, ge=0.0, le=1.0)


class OrderBasedRiskParameters(BaseModel):
    """Risk parameters for order-based trading strategy."""

    model_config = ConfigDict(extra="forbid")

    self_proportionality_multiplier: float = Field(default=1.0, gt=0.0)
    open_on_low_pnl: OpenOnLowPnl = Field(default_factory=OpenOnLowPnl)


class AutoBucketConfig(BaseModel):
    """Auto-detection bucket configuration."""

    model_config = ConfigDict(extra="forbid")

    type: Literal["auto"]
    ratio_threshold: float = Field(default=1000.0, ge=0.0)
    wide_bucket_percent: float = Field(default=0.01, gt=0.0, le=0.01)
    narrow_bucket_percent: float = Field(default=0.0001, ge=0.0, le=0.01)
    pricing_strategy: Literal["vwap", "aggressive"] = "vwap"


class ManualBucketConfig(BaseModel):
    """Fixed-width bucket configuration."""

    model_config = ConfigDict(extra="forbid")

    type: Literal["manual"]
    width_percent: float = Field(..., ge=0.0, le=1.0)
    pricing_strategy: Literal["vwap", "aggressive"] = "vwap"


BucketConfig = Annotated[
    AutoBucketConfig | ManualBucketConfig,
    Field(discriminator="type"),
]


class PositionBasedStrategy(BaseModel):
    """Position-based trading strategy (no bucket config required)."""

    model_config = ConfigDict(extra="forbid")

    type: Literal["position_based"]


class OrderBasedStrategy(BaseModel):
    """Order-based trading strategy with required bucket config."""

    model_config = ConfigDict(extra="forbid")

    type: Literal["order_based"]
    risk_parameters: OrderBasedRiskParameters = Field(default_factory=OrderBasedRiskParameters)
    bucket_config: BucketConfig


TradingStrategy = Annotated[
    PositionBasedStrategy | OrderBasedStrategy,
    Field(discriminator="type"),
]


class TraderSettings(BaseModel):
    """Trading strategy container."""

    model_config = ConfigDict(extra="forbid")

    trading_strategy: TradingStrategy


class TraderConfigSchema(BaseModel):
    """Complete trader configuration schema."""

    model_config = ConfigDict(extra="forbid")

    provider_settings: ProviderSettings
    trader_settings: TraderSettings


# Update-specific schemas where self_account.address is optional
# (auto-filled from trader's wallet_address in the service layer)


class SelfAccountUpdate(BaseModel):
    """Self account for updates - address is optional."""

    model_config = ConfigDict(extra="forbid")

    address: str | None = Field(default=None, pattern=r"^0x[a-fA-F0-9]{40}$")
    is_sub: bool = False


class ProviderSettingsUpdate(BaseModel):
    """Exchange and account configuration for updates."""

    model_config = ConfigDict(extra="forbid")

    exchange: Literal["hyperliquid"] = "hyperliquid"
    network: Literal["mainnet", "testnet"]
    self_account: SelfAccountUpdate = Field(default_factory=SelfAccountUpdate)
    copy_account: CopyAccount
    slippage_bps: int = Field(default=200, ge=0, le=1000)
    risk_parameters: ProviderRiskParameters


class TraderConfigUpdateSchema(BaseModel):
    """Trader configuration for updates - self_account.address is optional."""

    model_config = ConfigDict(extra="forbid")

    provider_settings: ProviderSettingsUpdate
    trader_settings: TraderSettings

File: web/src/lib/schemas/trader-config.ts
import { z } from "zod";

const ethereumAddressRegex = /^0x[a-fA-F0-9]{40}$/;

export const selfAccountSchema = z.object({
  address: z.string().regex(ethereumAddressRegex, "Invalid Ethereum address"),
  is_sub: z.boolean().default(false),
});

// For form validation, self_account.address can be empty (auto-filled from wallet_address)
export const selfAccountFormSchema = z.object({
  address: z.string().default(""),
  is_sub: z.boolean().default(false),
});

export const copyAccountSchema = z.object({
  address: z.string().regex(ethereumAddressRegex, "Invalid Ethereum address"),
});

export const providerSettingsSchema = z.object({
  exchange: z.literal("hyperliquid").default("hyperliquid"),
  network: z.enum(["mainnet", "testnet"]),
  self_account: selfAccountSchema,
  copy_account: copyAccountSchema,
  slippage_bps: z.number().int().min(0).max(1000).default(200),
  risk_parameters: z.object({
    allowed_assets: z.union([z.literal("*"), z.array(z.string()).min(1)]),
    blocked_assets: z.array(z.string()).default([]),
    max_leverage: z.number().int().min(1).max(50).nullable().optional(),
  }),
});

// For form validation, uses relaxed self_account validation
export const providerSettingsFormSchema = z.object({
  exchange: z.literal("hyperliquid").default("hyperliquid"),
  network: z.enum(["mainnet", "testnet"]),
  self_account: selfAccountFormSchema,
  copy_account: copyAccountSchema,
  slippage_bps: z.number().int().min(0).max(1000).default(200),
  risk_parameters: z.object({
    allowed_assets: z.union([z.literal("*"), z.array(z.string()).min(1)]),
    blocked_assets: z.array(z.string()).default([]),
    max_leverage: z.number().int().min(1).max(50).nullable().optional(),
  }),
});

export const openOnLowPnlSchema = z.object({
  enabled: z.boolean().default(false),
  max_pnl: z.number().min(0).max(1).default(0),
});

export const riskParametersSchema = z.object({
  self_proportionality_multiplier: z.number().min(0.01).max(10).default(1.0),
  open_on_low_pnl: openOnLowPnlSchema.default({}),
});

export const manualBucketSchema = z.object({
  type: z.literal("manual"),
  width_percent: z.number().min(0).max(1),
  pricing_strategy: z.enum(["vwap", "aggressive"]).default("vwap"),
});

export const autoBucketSchema = z.object({
  type: z.literal("auto"),
  ratio_threshold: z.number().min(0).default(1000),
  wide_bucket_percent: z.number().gt(0).max(0.01).default(0.01),
  narrow_bucket_percent: z.number().min(0).max(0.01).default(0.0001),
  pricing_strategy: z.enum(["vwap", "aggressive"]).default("vwap"),
});

export const bucketConfigSchema = z.discriminatedUnion("type", [
  autoBucketSchema,
  manualBucketSchema,
]);

export const tradingStrategySchema = z.object({
  type: z.literal("order_based").default("order_based"),
  risk_parameters: riskParametersSchema.default({}),
  bucket_config: bucketConfigSchema.default({ type: "auto", pricing_strategy: "vwap" }),
});

export const traderSettingsSchema = z.object({
  trading_strategy: tradingStrategySchema.default({}),
});

export const traderConfigSchema = z.object({
  provider_settings: providerSettingsSchema,
  trader_settings: traderSettingsSchema,
});

// Form-specific config schema with relaxed self_account validation
export const traderConfigFormSchema = z.object({
  provider_settings: providerSettingsFormSchema,
  trader_settings: traderSettingsSchema,
});

export type TraderConfig = z.infer<typeof traderConfigSchema>;
export type ProviderSettings = z.infer<typeof providerSettingsSchema>;
export type TraderSettings = z.infer<typeof traderSettingsSchema>;
export type TradingStrategy = z.infer<typeof tradingStrategySchema>;
export type RiskParameters = z.infer<typeof riskParametersSchema>;
export type BucketConfig = z.infer<typeof bucketConfigSchema>;

// Form-specific schema that includes wallet credentials
// Uses relaxed validation for self_account.address (auto-filled from wallet_address)
export const createTraderFormSchema = z
  .object({
    wallet_address: z
      .string()
      .regex(ethereumAddressRegex, "Invalid Ethereum address"),
    private_key: z
      .string()
      .regex(/^0x[a-fA-F0-9]{64}$/, "Invalid private key format"),
    name: z.string().max(50).optional(),
    description: z.string().max(255).optional(),
    config: traderConfigFormSchema,
  })
  .refine(
    (data) =>
      data.config.provider_settings.copy_account.address.toLowerCase() !==
      data.wallet_address.toLowerCase(),
    {
      message: "Copy account cannot be the same as your wallet",
      path: ["config", "provider_settings", "copy_account", "address"],
    }
  );

export type CreateTraderForm = z.infer<typeof createTraderFormSchema>;

// Edit form schema - only validates config, wallet_address and private_key are optional
export const editTraderFormSchema = z.object({
  wallet_address: z.string().optional(),
  private_key: z.string().optional(),
  name: z.string().max(50).optional(),
  description: z.string().max(255).optional(),
  config: traderConfigFormSchema,
});

export type EditTraderForm = z.infer<typeof editTraderFormSchema>;
